README.md
# Performance Audit Plugin for Matomo

[![Stable Version](https://img.shields.io/github/v/release/DevDavido/performance-audit-plugin)](https://github.com/DevDavido/performance-audit-plugin/releases)
[![GPL Licensed](https://img.shields.io/github/license/DevDavido/performance-audit-plugin?color=yellow)](LICENSE.md)
[![Tests Badge](https://img.shields.io/github/actions/workflow/status/DevDavido/performance-audit-plugin/tests.yml)](https://github.com/DevDavido/performance-audit-plugin/actions/workflows/tests.yml)

## Description
Daily performance audits of all your sites in Matomo for the following metrics based on Google Lighthouse:
- First Contentful Paint
- Speed Index
- Largest Contentful Paint
- Time To Interactive
- Total Blocking Time
- Cumulative Layout Shift
- Overall Score

Continuously monitor those Web Vital metrics over time, allowing detection of underlying problems before they have an adverse effect for users or simply track changes made to the web application, allowing you to establish a baseline for comparison too.

**Note:** Requires minimum NPM v10.x (part of Node.js 20.x LTS) to be installed on your server, otherwise plugin cannot be activated.

## Support me
If you installed this plugin and it was useful for you or your business, please don't hesitate to make a donation, I would highly appreciate it. Thank you!

<a href="https://www.buymeacoffee.com/devdavido" target="_blank"><img src="https://github.com/DevDavido/performance-audit-plugin/assets/997605/0c4e6a4c-9852-4410-a153-fa379fa7aa17" width="180" alt="Buy me a coffee"></a>

## Installation
Preferably you install [this plugin](https://plugins.matomo.org/PerformanceAudit) automatically by:
1. Login to Matomo as Super User
2. Navigate in your Matomo Installation to `Administration` › `Marketplace` (categorized under `Platform`)
3. Search for `Performance Audit`
4. Click on `Install` and wait up to 5 minutes for the installation to finish

You can also install this plugin manually by:
1. [downloading the latest plugin zip file](https://github.com/DevDavido/performance-audit-plugin/releases/latest)
2. Login to Matomo as Super User
3. Navigate in your Matomo Installation to `Administration` › `Marketplace` (categorized under `Platform`)
4. Click on `upload a Plugin` and upload the zip file from step 1
5. Activate this plugin `PerformanceAudit` and wait up to 5 minutes for the process to finish

Note: If plugin upload is disabled, enable it in your `config/config.ini.php` like that:
```ini
[General]
enable_plugin_upload = 1
```
If any errors occur during activation, please follow the instruction or information of the error message.

## Minimum requirements
- Matomo 5.0
- PHP 7.2.5
- NPM v10.x (part of [Node.js](https://nodejs.org/en/download/) 20.x LTS) to be installed on your server, otherwise plugin cannot be activated.
- If you use a UNIX based OS [these dependencies are required](https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#chrome-doesnt-launch-on-linux)

## Impact on server
Please keep in mind that Performance Audit uses Lighthouse internally in order to perform performance audits in the background for every Matomo tracked URL within the last 1 month.
If the Matomo installation has more than 10,000 tracked unique URLs within the last 1 month, it's suggested to either reduce the number of performed audit iterations in the website settings,
group URLs in website settings or increase the server resources. For Matomo installations with over 100,000 tracked unique URLs within the last 1 month,
this plugin architecture is rather unsuitable and therefor not recommended.

## Screenshots
### Dashboard
![Dashboard](/screenshots/Dashboard.png?raw=true)

### Overall Scores
![Overall Scores](/screenshots/OverallScores.png?raw=true)

### First Contentful Paint
![First Contentful Paint](/screenshots/FirstContentfulPaint.png?raw=true)

For more screenshots, check out the [screenshot overview](/screenshots/OVERVIEW.md).

## Testing
Run the integration and unit tests with:

```shell
./console tests:run PerformanceAudit
```
## Changelog
Please see the [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contact
If you have any questions or inquiries, you can contact `github{at}diskoboss{døt}de`.

## Security
If you discover any security related issues, please contact `github{at}diskoboss{døt}de` instead of using the issue tracker.

Please note that the performance audits will be performed with Chromium which renders Matomo-tracked web pages on your server without the sandbox mode of Chromium. You can find more information about this possible issue on the [lighthouse-cli repository](https://github.com/GoogleChrome/lighthouse-ci/tree/master/docs/recipes/docker-client#--no-sandbox-issues-explained).

## License
Licensed under the [GPLv3 License](LICENSE.md).


CHANGELOG.md
# Changelog
## 3.1.0
- Added: Search for Node executable additionally in paths of `PATH` environment variable
- Improved: Provide directory search path in error message if executable cannot be found

## 3.0.0
- Release stable version for Matomo 5
- Fixed: CORS template issue

## 3.0.0-Beta2
- Fixed: Issue with migrated Request handling

## 3.0.0-Beta1
- Added: Support for Matomo 5 🎉
- Updated: Bumped the minimum Node version to 20.x LTS for this new major plugin version
- Improved: Migrated some deprecated Matomo Plugin code

## 2.1.0
- Added: Site setting to group URLs by anchor which removes auditing duplicates of URLs which only differ in their anchors
- Added: Support to find latest Node executable in Plesk environment
- Added: Initial (partial) translation available for Albanian, Basque, Bokmål, Brazilian Portuguese, Bulgarian, Catalan, Chinese (Traditional), Dutch, French, German, Greek, Indonesian, Italian, Japanese, Portuguese, Spanish, Swedish, Turkish, Ukrainian
- Updated: Bumped `symfony/process`, `ducks-project/spl-types`, `symfony/polyfill-mbstring` and `friendsofphp/php-cs-fixer` dependency version
- Updated donation links

## 2.0.0
- Release stable version for Matomo 4
- Updated: Bumped `symfony/process` and `symfony/polyfill-mbstring` dependency version

## 2.0.0-Beta3
- Improved: Merged changes from 1.1.4 into 2.0.0-Beta2

## 2.0.0-Beta2
- Improved: Merged changes from 1.0.8 into 2.0.0-Beta2

## 2.0.0-Beta1
- Added: Support for Matomo 4 🎉
- Updated: Bumped the minimum PHP version to 7.2.5 for this new major plugin version, just as Matomo 4 itself

## 1.1.4
- Improved: Exception handling for failed audit due to too many requests response
- Fixed: One performance audit setting has been displayed after disabling audits for the site in settings
- Fixed: Disabled performance audits for site renders dashboard empty

## 1.1.3
- Added: Possibility to set extended audit timeout for each site
- Improved: Minor internal refactoring for site settings
- Updated: Bumped `symfony/process` dependency version

## 1.1.2
- Added: Possibility to enable or disable audit for each site
- Improved: Set timeout for installation process to 5 minutes

## 1.1.1
- Added: Scheduled weekly clearing of task running flag (in case of unexpected audit cancellation)
- Added: Console command to clear task running flag
- Fixed: Avoid race condition in case of which audit was called multiple times simultaneously
- Improved: Garbage collection after each audit
- Improved: Certain timeout/runtime exceptions don't stop following page audits anymore
- Improved: Set timeout for (audit) processes to 1 minute

## 1.1.0
- Fixed: Regular users cannot login anymore if plugin is activated
- Fixed: Renamed and fixed option to remove query strings from audited URLs which is now named group URLs and it doesn't throw SQL warnings anymore in certain (edge) cases
- Improved: Added security information for applied no sandbox mode of Chromium
- Improved: Increase database connection timeouts for longer running site audits
- Improved: Increase timeout for (audit) processes to 5 minutes
- Improved: Throw warning instead error for audited pages with HTTP 403 / 404 response
- Improved: Added FAQ entry for missing Chromium dependencies
- Improved: Error message during installation if directory permissions are incorrect

## 1.0.11
- Added: Option to remove query strings from audited URLs
- Improved: Small refactoring in settings

## 1.0.10
- Fixed: Issue with previous release as version number was incompatible with plugin marketplace

## 1.0.9
- Fixed: Schedule Reports and mobile app threw exception due to missing name attribute

## 1.0.8
- Improved: Refactored plugin base class, additionally removed now unneeded update classes
- Fixed: Updates could remove Node dependencies which now get reinstalled (regression bug)
- Fixed: Plugin cleanup now removes also symlinked directories

## 1.0.7
- Improved: Plugin check is now also running pre-checks first
- Improved: Plugin check is now independent of regular audit flow
- Improved: Wording in README text regarding tests
- Fixed: Updates could remove Node dependencies which now get reinstalled

## 1.0.6
- Fixed: Correct file require bug from previous release

## 1.0.5
- Added: Possibility to check plugin audit functionality

## 1.0.4
- Fixed: Switched to require `piwik` instead `matomo` in `plugin.json` for Matomo 3.x compatibility
- Improved: `plugin.json` plugin description

## 1.0.3
- Added: PerformanceAudit is live on the Matomo plugin marketplace now 🎉
- Added: Donation information in `plugin.json`
- Fixed: License name value in `plugin.json` was rewritten to a supported name value by Matomo Marketplace

## 1.0.2
- Fixed: Regression bug which made only first website get audited
- Improved: Logging of exceptions for audits
- Updated: Bumped `symfony/polyfill-mbstring` dependency version

## 1.0.1
- Fixed: Made sure tasks aren't executed concurrently
- Fixed: composer.json is compatible with Composer v2 now

## 1.0
- Initial release


Lighthouse.php
<?php

namespace Piwik\Plugins\PerformanceAudit;

require PIWIK_INCLUDE_PATH . '/plugins/PerformanceAudit/vendor/autoload.php';

use Dzava\Lighthouse\Lighthouse as BaseLighthouse;
use Piwik\Plugins\PerformanceAudit\Exceptions\AuditFailedAuthoriseRefusedException;
use Piwik\Plugins\PerformanceAudit\Exceptions\AuditFailedException;
use Piwik\Plugins\PerformanceAudit\Exceptions\AuditFailedMethodNotAllowedException;
use Piwik\Plugins\PerformanceAudit\Exceptions\AuditFailedNotFoundException;
use Piwik\Plugins\PerformanceAudit\Exceptions\AuditFailedTooManyRequestsException;
use Piwik\Plugins\PerformanceAudit\Exceptions\DependencyMissingException;
use Piwik\Plugins\PerformanceAudit\Exceptions\DependencyUnexpectedResultException;
use Symfony\Component\Process\Exception\ProcessSignaledException;
use Symfony\Component\Process\Exception\ProcessTimedOutException;
use Symfony\Component\Process\Exception\RuntimeException;
use UnexpectedValueException;

class Lighthouse extends BaseLighthouse
{
    protected $audits = [];

    /**
     * Setup and find lighthouse executable.
     *
     * @throws DependencyMissingException
     */
    public function __construct()
    {
        parent::__construct();

        $this->setChromePath($this->getChromePath());
        $this->setLighthousePath(ExecutableFinder::search('lighthouse'));
        $this->setTimeout(60);
    }

    /**
     * Perform audit of URL.
     *
     * @param string $url
     * @return string
     * @throws AuditFailedAuthoriseRefusedException|AuditFailedNotFoundException
     * @throws AuditFailedMethodNotAllowedException|AuditFailedException
     * @throws RuntimeException|ProcessTimedOutException|ProcessSignaledException
     */
    public function audit($url)
    {
        $command = $this->getCommand($url);
        $env = ['CHROME_PATH' => $this->getChromePath()];
        $process = new Process($command, $env);

        $process->setTimeout($this->timeout)->run();

        if (!$process->isSuccessful()) {
            $this->throwAuditException($url, $process->getErrorOutput());
        }

        return $process->getOutput();
    }

    /**
     * Create command by building config if needed and merging options.
     *
     * @param string $url
     * @return array
     */
    public function getCommand($url)
    {
        if ($this->configPath === null || $this->config !== null) {
            $this->buildConfig();
        }

        $command = array_merge([
            $this->lighthousePath,
            $this->outputFormat,
            $this->headers,
            '--quiet',
            '--config-path=' . $this->configPath,
            $url,
        ], $this->processOptions());

        return array_filter($command);
    }

    /**
     * Returns Chrome path from bundled browser of puppeteer.
     *
     * @return string
     * @throws DependencyUnexpectedResultException
     */
    public function getChromePath()
    {
        if (!is_null($this->chromePath)) {
            return $this->chromePath;
        }

        $jsFilePath = realpath(__DIR__ . DIRECTORY_SEPARATOR . 'ChromePath.js');
        $npmPath = ExecutableFinder::search('node');
        $process = new Process([$npmPath, '--no-warnings', $jsFilePath]);
        $process->run();

        if (!$process->isSuccessful()) {
            throw new DependencyUnexpectedResultException('NPM has the following unexpected output: ' . PHP_EOL . $process->getErrorOutput());
        }

        return trim($process->getOutput());
    }

    /**
     * Set emulated device.
     *
     * @param string $device
     * @return self
     */
    public function setEmulatedDevice(string $device)
    {
        if (!in_array($device, ['desktop', 'mobile'])) {
            throw new UnexpectedValueException('Lighthouse emulated device has no valid value.');
        }
        $this->setOption('--emulated-form-factor', $device);

        return $this;
    }

    /**
     * Set Chrome path.
     *
     * @param string $path
     * @return self
     */
    public function setChromePath($path)
    {
        $this->chromePath = $path;

        return $this;
    }

    /**
     * Disable all device emulation.
     *
     * @return self
     */
    public function disableDeviceEmulation()
    {
        $this->setOption('--emulated-form-factor', 'none');
        $this->setOption('--throttling-method', 'provided');

        return $this;
    }

    /**
     * Enable audit.
     *
     * @param string $audit
     * @return self
     */
    public function enableAudit($audit)
    {
        $this->setAudit($audit, true);

        return $this;
    }

    /**
     * Creates the config file used during the audit.
     *
     * @return self
     */
    protected function buildConfig()
    {
        $config = tmpfile();
        $this->withConfig(stream_get_meta_data($config)['uri']);
        $this->config = $config;

        $r = 'module.exports = ' . json_encode([
                'extends' => 'lighthouse:default',
                'settings' => [
                    'onlyCategories' => $this->categories,
                    'onlyAudits' => $this->audits,
                ],
            ]);
        fwrite($config, $r);

        return $this;
    }

    /**
     * Enable or disable a audit.
     *
     * @param string $audit
     * @param bool $enable
     * @return self
     */
    protected function setAudit($audit, $enable)
    {
        $index = array_search($audit, $this->audits);

        if ($index !== false) {
            if ($enable == false) {
                unset($this->audits[$index]);
            }
        } elseif ($enable) {
            $this->audits[] = $audit;
        }

        return $this;
    }

    /**
     * Throw an corresponding exception if audit fails.
     *
     * @param string $url
     * @param string $errorOutput
     * @return void
     * @throws AuditFailedAuthoriseRefusedException|AuditFailedNotFoundException
     * @throws AuditFailedMethodNotAllowedException|AuditFailedException
     * @throws AuditFailedTooManyRequestsException
     */
    protected function throwAuditException($url, $errorOutput)
    {
        if (stristr($errorOutput, 'Status code: 403')) {
            throw new AuditFailedAuthoriseRefusedException($url, $errorOutput);
        } elseif (stristr($errorOutput, 'Status code: 404')) {
            throw new AuditFailedNotFoundException($url, $errorOutput);
        } elseif (stristr($errorOutput, 'Status code: 405')) {
            throw new AuditFailedMethodNotAllowedException($url, $errorOutput);
        } elseif (stristr($errorOutput, 'Status code: 429')) {
            throw new AuditFailedTooManyRequestsException($url, $errorOutput);
        }

        throw new AuditFailedException($url, $errorOutput);
    }
}
